scanner/api/sink.h

```diff
@@ -0,0 +1,181 @@
+
+  /**
+   * @brief Runs the Sink to write elements.
+   *
+   * @param input_columns
+   *        sink input, vector of elements produced by an Op
+   *
+   */
+  virtual void write(const BatchedElements& input_columns) = 0;
+
+  /**
+   * @brief Called when the Sink has received the end of a contiguous sequence
+   *        of data items (for example, when the last element of a stream is
+   *        provided to the Sink).
+   *
+   */
+  virtual void finished() {};
+
+  /**
+   * @brief For internal use
+   **/
+  virtual void set_profiler(Profiler* profiler) { profiler_ = profiler; }
+
+  /**
+   * The profiler allows an op to save profiling data for later
+   * visualization. It is not guaranteed to be non-null, so check before use.
+   */
+  Profiler* profiler_ = nullptr;
+};
+
+///////////////////////////////////////////////////////////////////////////////
+/// Implementation Details
+namespace internal {
+
+class SinkBuilder;
+
+using SinkConstructor =
+    std::function<Sink*(const SinkConfig& config)>;
+
+class SinkRegistration {
+ public:
+  SinkRegistration(const SinkBuilder& builder);
+};
+
+class SinkBuilder {
+ public:
+  friend class SinkRegistration;
+
+  SinkBuilder(const std::string& name, SinkConstructor constructor)
+    : name_(name),
+      constructor_(constructor),
+      variadic_inputs_(false),
+      per_element_output_(false),
+      entire_stream_output_(false) {}
+
+  SinkBuilder& variadic_inputs() {
+    if (input_columns_.size() > 0) {
+      LOG(FATAL) << "Sink " << name_ << " cannot have both fixed and variadic "
+                 << "inputs";
+    }
+    variadic_inputs_ = true;
+    return *this;
+  }
+
+  SinkBuilder& input(const std::string& name,
+                   ColumnType type = ColumnType::Other) {
+    if (variadic_inputs_) {
+      LOG(FATAL) << "Sink " << name_ << " cannot have both fixed and variadic "
+                 << "inputs";
+    }
+    input_columns_.push_back(std::make_tuple(name, type));
+    return *this;
+  }
+
+  SinkBuilder& frame_input(const std::string& name) {
+    return input(name, ColumnType::Video);
+  }
+
+  SinkBuilder& per_element_output() {
+    if (entire_stream_output_) {
+      LOG(FATAL) << "Sink " << name_
+                 << " cannot specify both per element and entire stream output";
+    }
+    per_element_output_ = true;
+    return *this;
+  }
+
+  SinkBuilder& entire_stream_output() {
+    LOG(FATAL) << "Entire stream output is not implemented yet.";
+
+    if (per_element_output_) {
+      LOG(FATAL) << "Sink " << name_
+                 << " cannot specify both per element and entire stream output";
+    }
+    entire_stream_output_ = true;
+    return *this;
+  }
+
+ private:
+  std::string name_;
+  SinkConstructor constructor_;
+  bool variadic_inputs_;
+  std::vector<std::tuple<std::string, ColumnType>> input_columns_;
+  bool per_element_output_;
+  bool entire_stream_output_;
+};
+}
+
+#define REGISTER_SINK(name__, sink__) \
```

Comment:
If we get enough dynamic registries, maybe we'll need a `REGISTER_REGISTRY` to help us make generic infrastructure for registration ;D

Reply:
Yeah, I was thinking about that :P